### PR TITLE
[FrameworkBundle] Fix service reset between tests

### PR DESCRIPTION
**Pull request by HypeMC**
| Q             | A
| ------------- | ---
| Branch?       | 5.4
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Issues        | -
| License       | MIT

Currently, not all services are reset between tests. If a service uses the kernel.reset tag but does not implement the ResetInterface, it never gets reset.

### CODE REVIEW

File: src/Symfony/Bundle/FrameworkBundle/Tests/Functional/Bundle/TestBundle/TestServiceContainer/ResettableService.php [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Bundle\FrameworkBundle\Tests\Functional\Bundle\TestBundle\TestServiceContainer;
+
+class ResettableService
+{
+    private $count = 0;
```
Comment (OskarStark) [inline src/Symfony/Bundle/FrameworkBundle/Tests/Functional/Bundle/TestBundle/TestServiceContainer/ResettableService.php:16]:
```suggestion
    private int $count = 0;
```
nitpick

File: src/Symfony/Bundle/FrameworkBundle/Tests/Functional/app/TestServiceContainer/services.yml [hunk 2]
```diff
@@ -13,3 +13,8 @@ services:
         arguments:
             - '@Symfony\Bundle\FrameworkBundle\Tests\Functional\Bundle\TestBundle\TestServiceContainer\NonPublicService'
             - '@Symfony\Bundle\FrameworkBundle\Tests\Functional\Bundle\TestBundle\TestServiceContainer\PrivateService'
+
+    Symfony\Bundle\FrameworkBundle\Tests\Functional\Bundle\TestBundle\TestServiceContainer\ResettableService:
+        public: true
+        tags:
+            -   kernel.reset: { method: 'myCustomName' }
```
Comment (OskarStark) [inline src/Symfony/Bundle/FrameworkBundle/Tests/Functional/app/TestServiceContainer/services.yml]:
```suggestion
            - kernel.reset: { method: 'myCustomName' }
```

**Review by nicolas-grekas — APPROVED** [review]

File: src/Symfony/Bundle/FrameworkBundle/Tests/Functional/app/TestServiceContainer/services.yml (hunk 2 above)
Comment (HypeMC) [reply src/Symfony/Bundle/FrameworkBundle/Tests/Functional/app/TestServiceContainer/services.yml]: Done

File: src/Symfony/Bundle/FrameworkBundle/Tests/Functional/Bundle/TestBundle/TestServiceContainer/ResettableService.php (hunk 1 above)
Comment (HypeMC) [reply src/Symfony/Bundle/FrameworkBundle/Tests/Functional/Bundle/TestBundle/TestServiceContainer/ResettableService.php:16]: Can't, requires PHP 7.4, Symfony 5.4 is 7.2 compatible.

File: src/Symfony/Bundle/FrameworkBundle/Test/KernelTestCase.php [hunk 3]
```diff
@@ -160,7 +160,9 @@ protected static function ensureKernelShutdown()
             static::$kernel->shutdown();
             static::$booted = false;
 
-            if ($container instanceof ResetInterface) {
+            if ($container->has('services_resetter')) {
```
Comment (xabbuh) [inline src/Symfony/Bundle/FrameworkBundle/Test/KernelTestCase.php:163]: Does the added rest really reach this part? The container still implements the `ResetInterface`, doesn’t it?

Comment (HypeMC) [reply src/Symfony/Bundle/FrameworkBundle/Test/KernelTestCase.php:163]: Sorry, I'm not sure I understand the question.

Comment (xabbuh) [reply src/Symfony/Bundle/FrameworkBundle/Test/KernelTestCase.php:163]: Let me rephrase it. I would expect the added test to pass even without the change here.

Comment (HypeMC) [reply src/Symfony/Bundle/FrameworkBundle/Test/KernelTestCase.php:163]: But it doesn't. `ResettableService` doesn't implement `ResetInterface`, which is how the `Container::reset()` method works: https://github.com/symfony/symfony/blob/fea8f98676e705d696307708f599ad096c633406/src/Symfony/Component/DependencyInjection/Container.php#L299-L313

The `services_resetter` on the other hand takes into account services without the interface:  
https://github.com/symfony/symfony/blob/fea8f98676e705d696307708f599ad096c633406/src/Symfony/Component/HttpKernel/DependencyInjection/ServicesResetter.php#L39-L50

File: src/Symfony/Bundle/FrameworkBundle/Test/KernelTestCase.php [hunk 4]
```diff
@@ -160,7 +160,9 @@ protected static function ensureKernelShutdown()
...
+                $container->get('services_resetter')->reset();
+            } elseif ($container instanceof ResetInterface) {
                 $container->reset();
```
Comment (stof) [inline src/Symfony/Bundle/FrameworkBundle/Test/KernelTestCase.php:169]: `$container->reset()` must still be called, to reset the instances hold by the container itself (see the code of line 302).

Btw, if you ensure that the `services_resetter` service is instantiated, the container reset call will then call it (as ServiceResetter implements the ResetInterface).

The reason it was not called is because `Container::reset` only resets *instantiated* resettable services

Comment (HypeMC) [reply src/Symfony/Bundle/FrameworkBundle/Test/KernelTestCase.php:169]: @stof Fixed, see https://github.com/symfony/symfony/compare/549ada5257c771da92d51aff76e224e1ac5d31f4..5491aa54a3205acfaf5ba62c134bddc26d794a36

Initially, I wanted to just instantiate the `services_resetter` and let `Container::reset()` call reset on it, but that didn’t work because the `services_resetter` only resets instantiated services. Because of line 302, by the time it’s called, there aren’t any instantiated services left.

The downside of the current approach is that some services will get reset twice, but I don’t see a better solution at the moment.

Comment (stof) [reply src/Symfony/Bundle/FrameworkBundle/Test/KernelTestCase.php:169]: ah indeed. The container resets itself before resetting the services.

I think this might actually be an issue if the `reset` method of a service triggers the instantiation of some other service. I think the resetting of the container state should probably be moved after the loop.

Comment (HypeMC) [reply src/Symfony/Bundle/FrameworkBundle/Test/KernelTestCase.php:169]: @stof Done https://github.com/symfony/symfony/compare/5491aa54a3205acfaf5ba62c134bddc26d794a36..53d0f56be6a315055cb3ceef09a7f6bcb233a120

**Review by stof — APPROVED** [review]

**Comment by xabbuh** [comment]
Isn’t the test failure on PHP 8.2 related?

**Comment by HypeMC** [comment]
> Isn’t the test failure on PHP 8.2 related?

@xabbuh Yes, but that's high-deps and if I'm not mistaking, that'll get resolved when merged upstream.

File: src/Symfony/Bundle/FrameworkBundle/Test/KernelTestCase.php (hunk 3 above)
Comment (xabbuh) [reply src/Symfony/Bundle/FrameworkBundle/Test/KernelTestCase.php:163]: thanks for the clarification 👍

**Comment by nicolas-grekas** [comment]
Thank you @HypeMC.
